### ISSUE TITLE
[Feature Request] Enable Link-Time Optimization (LTO)

### ISSUE THREAD

**Issue by zamazan4ik** [glzr-io/zebar#128]
### Describe the problem/motivation

Motivation: improve the binary size and, possibly, gain a bit of additional performance (but the binary size is the main motivation in this case).

### Describe the solution you'd like

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48).

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

I have made quick tests on my local setup (Fedora 40, Rustc 1.82): enabling `lto = true` for the Release profile reduces the binary size from 28 Mib to 20 Mib.

### Alternatives considered

Leave the things as is

**Comment by lars-berger** [comment]
> From 49MB to 16MB on Windows 11 using target x86_64-pc-windows-gnu, could probably be improved with x86_64-pc-windows-msvc but I don't want to pollute my windows setting up rust etc.

Ah interesting! Did a test on non-WSL a while back and the size difference was quite negligible (about 10-15% if I remember correctly). It was a significant increase in our build times though, so I decided not to pursue it further.

A 49MB -> 16MB difference would be worthwhile though. I'm not very familiar with the `x86_64-pc-windows-gnu` toolchain - do you know whether it'd be runnable on non-WSL systems?

**Comment by krishenriksen** [comment]
Didn't look into it, I just set up the environment in wsl -> ubuntu, with rust toolchain and pnpm, and added rustup target "x86_64-pc-windows-gnu"

**Comment by MarcosCosmos** [comment]
Low hanging fruit, happy to do this once we have it verified to be working across windows/linux/macos.
